plugins/rbac-backend/src/role-manager/role-manager.ts

```diff
@@ -227,9 +207,89 @@ export class BackstageRoleManager implements RoleManager {
     return newRole;
   }
 
+  private removeRole(name: string): void {
+    this.allRoles.delete(name);
+  }
+
   // parse the entity to find out if it is a user / group / or role
   private parseEntityKind(name: string): string {
     const parsed = name.split(':');
     return parsed[0];
   }
+
+  private nonCachedCheck(memo: AncestorSearchMemo, name2: string): boolean {
+    if (!memo.hasEntityRef(name2) && this.parseEntityKind(name2) === 'role') {
+      for (const [key, value] of this.allRoles.entries()) {
+        if (value.hasRole(name2, 1) && memo.hasEntityRef(key)) {
+          return true;
+        }
+      }
+    }
+    return memo.hasEntityRef(name2);
+  }
+
+  private detectCycleError(memo: AncestorSearchMemo, name1: string): boolean {
+    memo.debugNodesAndEdges(this.log, name1);
+    if (!memo.isAcyclic()) {
+      const cycles = memo.findCycles();
+
+      this.log.warn(
+        `Detected cycle dependencies in the Group graph: ${JSON.stringify(
+          cycles,
+        )}. Admin/(catalog owner) have to fix it to make RBAC permission evaluation correct for groups: ${JSON.stringify(
+          cycles,
+        )}`,
+      );
+
+      return false;
+    }
+    return true;
+  }
+
+  private async buildGraph(name1: string): Promise<AncestorSearchMemo> {
+    const memo = new AncestorSearchMemo(
+      name1,
+      this.tokenManager,
+      this.catalogApi,
+      this.catalogDBClient,
+    );
+    await memo.buildUserGraph(memo);
+
+    memo.debugNodesAndEdges(this.log, name1);
+    return memo;
+  }
+
+  private parseRoles(userAndParentGroups: string[]): string[] {
```

Comment:
Perhaps the name "matchRoles" would be better?

Reply:
Sounds good with me.